File: design_patterns/authenticated_users.py
class IAuthenticator:
  def authenticate(self, username: str, password: str) -> bool:
    pass

class Authenticator(IAuthenticator):
  def authenticate(self, username: str, password: str) -> bool:
    pass

class GoogleAuthenticator(IAuthenticator):
  def authenticate(self, username: str, password: str) -> bool:
    pass

class IUserRepository:
  def saveAuthenticatedUser(self, username: str) -> None:
    pass

class UserRepository(IUserRepository):
  def saveAuthenticatedUser(self, username: str) -> None:
    pass

class AuthController:
  def __init__(self, authenticator: IAuthenticator, userRepository: IUserRepository):
    self.authenticator = authenticator
    self.userRepository = userRepository

  def login(self, username: str, password: str) -> bool:
    if self.authenticator.authenticate(username, password):
      self.userRepository.saveAuthenticatedUser(username);
      return True
    else:
      return False

authenticator = Authenticator()
userRepository = UserRepository()
authController = AuthController(authenticator, userRepository)

File: design_patterns/authenticated_users.ts
interface IAuthenticator {
  authenticate(username: string, password: string): boolean;
}

class Authenticator implements IAuthenticator {
  authenticate(username: string, password: string): boolean {}
}

class GoogleAuthenticator implements IAuthenticator {
  authenticate(username: string, password: string): boolean {}
}

interface IUserRepository {
  saveAuthenticatedUser(username: string): void;
}

class UserRepository implements IUserRepository {
  saveAuthenticatedUser(username: string): void {}
}

class AuthController {
  private authenticator: IAuthenticator;
  private userRepository: IUserRepository;

  constructor(authenticator: IAuthenticator, userRepository: IUserRepository) {
    this.authenticator = authenticator;
    this.userRepository = userRepository;
  }

  login(username: string, password: string) {
    if (this.authenticator.authenticate(username, password)) {
      this.userRepository.saveAuthenticatedUser(username);
      return true;
    }
  }
}

const authenticator = new Authenticator();
const userRepository = new UserRepository();
const authController = new AuthController(authenticator, userRepository);
